### PR TITLE
JSDK-2666: Add remote participant reconnection example

### PR DESCRIPTION
**Pull request by PikaJoyce**
Creating a new example for remoteParticipant reconnection states. This app allows users to apply UI changes when a remoteParticipant is having network interruptions.
Screen Shot 2020-02-19 at 4 10 45 PM

Working demo 
Displays what the localParticipant will see (localState), as well as what the remoteParticipant will see when it's remoteParticipant (P1) reconnects/reconnecting
recstate

**Contributing to Twilio**

> All third-party contributors acknowledge that any contributions they provide will be made under the same open-source license that the open-source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

### CODE REVIEW

File: examples/remotereconnection/src/helpers.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
+'use strict';
+
+/**
+ * Listen to Room reconnection events and update the UI accordingly.
+ * @param {Room} room - The Room you have joined
+ * @param {function} updateRoomState - Updates the app UI with the new state
+ * @returns {void}
+ */
+
+ function remoteReconnectionUpdates (room, updateRemoteState) {
```
Comment (manjeshbhargav) [inline examples/remotereconnection/src/helpers.js]:
```js
function handleRemoteParticipantReconnectionUpdates(room, updateParticipantState) {
  room.on('participantReconnecting', participant => {
    updateParticipantState(participant.state);
  });

  room.on('participantReconnected', participant => {
    updateParticipantState(participant.state);
  });
}

function handleLocalParticipantReconnectionUpdates(room, updateParticipantState) {
  const localParticipant = room.localParticipant;

  localParticipant.on('reconnecting', () => {
    updateParticipantState(localParticipant.state);
  });

  localParticipant.on('reconnected', () => {
    updateParticipantState(localParticipant.state);
  });
} 
```

File: examples/remotereconnection/src/index.js [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+  const pre = document.querySelector('pre.language-javascript');
+  
+  pre.innerHTML = Prism.highlight(snippet, Prism.languages.javascript);
+  
+  // Get the credentials to connect to the Room.
+  const credsP1 = await getRoomCredentials();
+  const credsP2 = await getRoomCredentials();
+  
+  // Create room instance and name for participants to join.
+  const roomP1 = await Video.connect(credsP1.token);
```
Comment (manjeshbhargav) [inline examples/remotereconnection/src/index.js]:
```suggestion
  const roomP1 = await Video.connect(credsP1.token, { tracks: [] });
```

File: examples/remotereconnection/src/index.js [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+  const roomName = roomP1.name;
+  
+  // Connecting remote participants.
+  const roomP2 = await Video.connect(credsP2.token, {
+    name: roomName
+  });
+  
+  // Appends video/audio tracks when each participant is subscribed.
+  roomP1.on('trackSubscribed', track => {
+    if (track.isEnabled) {
```
Comment (manjeshbhargav) [inline examples/remotereconnection/src/index.js]:
```suggestion
```

File: examples/remotereconnection/src/index.js [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+      p1Media.appendChild(track.attach());
+    }
```
Comment (manjeshbhargav) [inline examples/remotereconnection/src/index.js]:
```suggestion
```

File: examples/remotereconnection/src/index.js [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+  
+  // Appends video/audio tracks when each participant is subscribed.
+  roomP1.on('trackSubscribed', track => {
+    if (track.isEnabled) {
+      p1Media.appendChild(track.attach());
+    } 
+  });
+  
+  // Simulate reconnection button functionalities, adding in region in order to extend reconnection time
+  P1simulateReconnection.onclick = () => {
```
Comment (manjeshbhargav) [inline examples/remotereconnection/src/index.js]:
```suggestion
  P2simulateReconnection.onclick = () => {
```

File: examples/remotereconnection/src/index.js [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+  });
+
+  // Local room listening on it's own reconnection state
+  roomP1.on('reconnecting', () => {
+    onRoomStateChange('p1', roomP1.state);
+  });
+
+  roomP1.on('reconnected', () => {
+    onRoomStateChange('p1', roomP1.state);
+  });
```
Comment (manjeshbhargav) [inline examples/remotereconnection/src/index.js]: This should be:
```js
handleRemoteParticipantReconnectionUpdates(roomP1, state => {
  onRoomStateChange('p2', state);
});

handleLocalParticipantReconnectionUpdates(roomP1, state => {
  onRoomStateChange('p1', state);
});
```

File: examples/remotereconnection/src/index.js (hunk 2 above)
Comment (PikaJoyce) [reply examples/remotereconnection/src/index.js]: Do we not want the local participant (P1) to show video/audio? Or should I make P2 to not publish tracks?

Comment (manjeshbhargav) [reply examples/remotereconnection/src/index.js]: @PikaJoyce Ideally you should be using only one of the Rooms to showcase both LocalParticipant and RemoteParticipant reconnection states (RoomP1 in this example). So, when you close the connection for RoomP2, RoomP1 emits a "participantReconnecting" event. When you close the connection for RoomP1, then RoomP1.localParticipant emits a "reconnecting" event.

**Comment by PikaJoyce** [comment]
Fixed!
Screen Shot 2020-02-21 at 6 02 14 PM

File: examples/remotereconnection/public/index.html [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+              Remote Participant Reconnection States
+            </h4>
+            <pre class="language-javascript"></pre>
+          </div>
+        </div>
+      </div>
+      <div class="col-sm-4">
+        <div class="card">
+          <div class="card-block">
+            <h4 class="card-title">Participant 1 : Local View</h4>
```
Comment (manjeshbhargav) [inline examples/remotereconnection/public/index.html]:
```suggestion
            <h4 class="card-title">Local View</h4>
```

File: examples/remotereconnection/public/index.html [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+            <h5> Local state is:</h5>
+            <div id="p1">
+              <div class="roomstate connected current">Connected</div>
+              <div class="roomstate reconnecting">Reconnecting</div>
+            </div>
+          </div>
+        </div>
+            <div class="card">
+              <div class="card-block">
+                <h4 class="card-title">Participant 2 : Remote View</h4>
```
Comment (manjeshbhargav) [inline examples/remotereconnection/public/index.html]:
```suggestion
                <h4 class="card-title">Remote View</h4>
```

File: examples/remotereconnection/public/index.html [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+          </div>
+        </div>
+      </div>
+      <div class="col-sm-4">
+        <div class="card">
+          <div class="card-block">
+            <h4 class="card-title">Participant 1 : Local View</h4>
+            <div id="p1-media"></div>
+            <button id="p1-simulate-reconnection" class="btn btn-primary btn-block">Simulate Reconnection</button>
+            <h5> Local state is:</h5>
```
Comment (manjeshbhargav) [inline examples/remotereconnection/public/index.html]:
```suggestion
            <h5>State</h5>
```

File: examples/remotereconnection/public/index.html [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+              <div class="roomstate connected current">Connected</div>
+              <div class="roomstate reconnecting">Reconnecting</div>
+            </div>
+          </div>
+        </div>
+            <div class="card">
+              <div class="card-block">
+                <h4 class="card-title">Participant 2 : Remote View</h4>
+                <button id="p2-simulate-reconnection" class="btn btn-primary btn-block">Simulate Reconnection</button>
+                <h5> Remote Participant State (Participant 1):</h5>
```
Comment (manjeshbhargav) [inline examples/remotereconnection/public/index.html]:
```suggestion
                <h5>State</h5>
```

File: examples/remotereconnection/src/helpers.js [hunk 11]
```diff
@@ -0,0 +1,33 @@
+'use strict';
+
+/**
+ * Listen to Room reconnection events and update the UI accordingly.
+ * @param {Room} room - The Room you have joined
+ * @param {function} updateRoomState - Updates the app UI with the new state
+ * @returns {void}
+ */
+
+ function handleRemoteParticipantReconnectionUpdates(room, updateParticipantState) {
```
Comment (manjeshbhargav) [inline examples/remotereconnection/src/helpers.js]:
```suggestion
function handleRemoteParticipantReconnectionUpdates(room, updateParticipantState) {
```

File: examples/remotereconnection/src/helpers.js [hunk 12]
```diff
@@ -0,0 +1,33 @@
+'use strict';
+
+/**
+ * Listen to Room reconnection events and update the UI accordingly.
+ * @param {Room} room - The Room you have joined
+ * @param {function} updateRoomState - Updates the app UI with the new state
+ * @returns {void}
+ */
+
```
Comment (manjeshbhargav) [inline examples/remotereconnection/src/helpers.js]:
```suggestion
```

File: examples/remotereconnection/src/helpers.js [hunk 13]
```diff
@@ -0,0 +1,33 @@
+'use strict';
+
+/**
+ * Listen to Room reconnection events and update the UI accordingly.
```
Comment (manjeshbhargav) [inline examples/remotereconnection/src/helpers.js]:
```suggestion
 * Listen to RemoteParticipant reconnection events and update the UI accordingly.
```

File: examples/remotereconnection/src/helpers.js [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+  room.on('participantReconnecting', participant => {
+    updateParticipantState(participant.state);
+  });
+
+  room.on('participantReconnected', participant => {
+    updateParticipantState(participant.state);
+  });
+}
+
+function handleLocalParticipantReconnectionUpdates(room, updateParticipantState) {
```
Comment (manjeshbhargav) [inline examples/remotereconnection/src/helpers.js]:
```suggestion
/**
 * Listen to LocalParticipant reconnection events and update the UI accordingly.
 * @param {Room} room - The Room you have joined
 * @param {function} updateRoomState - Updates the app UI with the new state
 * @returns {void}
 */
function handleLocalParticipantReconnectionUpdates(room, updateParticipantState) {
```

File: examples/remotereconnection/src/helpers.js [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+
+/**
+ * Listen to Room reconnection events and update the UI accordingly.
+ * @param {Room} room - The Room you have joined
+ * @param {function} updateRoomState - Updates the app UI with the new state
+ * @returns {void}
+ */
+
+ function handleRemoteParticipantReconnectionUpdates(room, updateParticipantState) {
+  room.on('participantReconnecting', participant => {
```
Comment (manjeshbhargav) [inline examples/remotereconnection/src/helpers.js]:
```suggestion
  room.on('participantReconnecting', function(participant) {
```

File: examples/remotereconnection/src/helpers.js [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+    updateParticipantState(participant.state);
+  });
+
+  room.on('participantReconnected', participant => {
```
Comment (manjeshbhargav) [inline examples/remotereconnection/src/helpers.js]:
```suggestion
  room.on('participantReconnected', function(participant) {
```

File: examples/remotereconnection/src/helpers.js [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+    updateParticipantState(participant.state);
+  });
+}
+
+function handleLocalParticipantReconnectionUpdates(room, updateParticipantState) {
+  const localParticipant = room.localParticipant;
+
+  localParticipant.on('reconnecting', () => {
```
Comment (manjeshbhargav) [inline examples/remotereconnection/src/helpers.js]:
```suggestion
  localParticipant.on('reconnecting', function() {
```

File: examples/remotereconnection/src/helpers.js [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+    updateParticipantState(localParticipant.state);
+  });
+
+  localParticipant.on('reconnected', () => {
```
Comment (manjeshbhargav) [inline examples/remotereconnection/src/helpers.js]:
```suggestion
  localParticipant.on('reconnected', function() {
```

File: examples/remotereconnection/src/index.js [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+    region: 'au1'
+  });
+  
+  // Appends video/audio tracks when each participant is subscribed.
+  roomP1.on('participantConnected', () => {
+    const localUser = roomP1.localParticipant;
+    getTracks(localUser).forEach(track => {
+      p1Media.appendChild(track.attach());
+    })
+  });
```
Comment (manjeshbhargav) [inline examples/remotereconnection/src/index.js]: @PikaJoyce You need not wait for the RemoteParticipant to join to attach the LocalParticipant's Tracks:
```suggestion
  getTracks(roomP1.localParticipant).forEach(track => {
    p1Media.appendChild(track.attach());
  });
```

Comment (manjeshbhargav) [reply examples/remotereconnection/src/index.js]: @PikaJoyce Also, can you attach the Tracks of the RemoteParticipant to the remote view as well?
```js
roomP1.on('trackSubscribed', track => {
  p2Media.appendChild(track.attach());
});
```
And just to be sure that you get this event, please move the above two code snippets before `roomP2` is created.

**Pull request merged**